rust/plugin-lib/src/state_cache.rs

```diff
@@ -181,12 +181,16 @@ impl<'a, S: Default + Clone> PluginCtx<'a, S> {
     fn do_update<P>(mut self, update: plugin_rpc::PluginUpdate, handler: &mut P) -> Value
         where P: Plugin<State = S>
     {
-        let plugin_rpc::PluginUpdate { delta, rev, .. } = update;
-        self.state.buf_size = delta.new_document_len();
+        let plugin_rpc::PluginUpdate { delta, new_len, rev, .. } = update;
+        self.state.buf_size = new_len;
         self.state.rev = rev;
-
-        self.update_line_cache(&delta);
-        self.update_chunk(&delta);
+        if let Some(ref delta) = delta {
+            self.update_line_cache(delta);
+            self.update_chunk(delta);
+        } else {
+            // if there's no delta (very large insert) we blow away everything
```

Comment:
I'd say "edit" rather than "insert" here, for precision, as there are other edits that may qualify.